daphne/src/hpke.rs

```diff
@@ -184,75 +133,39 @@ impl HpkeReceiverConfig {
-                >(&OpModeR::Base, &sk, &enc, info)?;
-                let pt = receiver.open(ciphertext, aad)?;
-                Ok(pt)
-            }
-            HpkeKemId::X25519HkdfSha256 => {
-                let sk = <kem::X25519HkdfSha256 as Kem>::PrivateKey::from_bytes(&self.secret_key)?;
-                let enc = <kem::X25519HkdfSha256 as Kem>::EncappedKey::from_bytes(enc)?;
-                let mut receiver = hpke::setup_receiver::<
-                    aead::AesGcm128,
-                    kdf::HkdfSha256,
-                    kem::X25519HkdfSha256,
-                >(&OpModeR::Base, &sk, &enc, info)?;
-                let pt = receiver.open(ciphertext, aad)?;
-                Ok(pt)
-            }
-            _ => {
-                // We should never get here.
-                unreachable!("Unsupported KEM ({:?})", self.config.kem_id);
-            }
-        }
+        let receiver: Hpke<ImplHpkeCrypto> =
+            check_suite(self.config.kem_id, self.config.kdf_id, self.config.aead_id)?;
+        let sk = HpkePrivateKey::new(self.secret_key.clone());
+        let mut ctx = receiver.setup_receiver(enc, &sk, info, None, None, None)?;
+        let plaintext = ctx.open(aad, ciphertext)?;
+        Ok(plaintext)
     }
 
     /// Generate and return a new HPKE receiver context given a HPKE config ID and HPKE KEM.
     pub fn gen(id: u8, kem_id: HpkeKemId) -> Self {
-        use hpke::Serializable;
-
-        match kem_id {
-            HpkeKemId::X25519HkdfSha256 => {
-                let (sk, pk) = kem::X25519HkdfSha256::gen_keypair(&mut rand::thread_rng());
-                HpkeReceiverConfig {
-                    config: HpkeConfig {
-                        id,
-                        kem_id,
-                        kdf_id: HpkeKdfId::HkdfSha256,
-                        aead_id: HpkeAeadId::Aes128Gcm,
-                        public_key: pk.to_bytes().to_vec(),
-                    },
-                    secret_key: sk.to_bytes().to_vec(),
-                }
-            }
-            HpkeKemId::P256HkdfSha256 => {
-                let (sk, pk) = kem::DhP256HkdfSha256::gen_keypair(&mut rand::thread_rng());
-                HpkeReceiverConfig {
-                    config: HpkeConfig {
-                        id,
-                        kem_id,
-                        kdf_id: HpkeKdfId::HkdfSha256,
-                        aead_id: HpkeAeadId::Aes128Gcm,
-                        public_key: pk.to_bytes().to_vec(),
-                    },
-                    secret_key: sk.to_bytes().to_vec(),
-                }
-            }
-            _ => {
-                // We should never get here.
-                unreachable!("Unsupported KEM ({:?})", kem_id);
-            }
+        let kem = match kem_id {
+            HpkeKemId::P256HkdfSha256 => KemAlgorithm::DhKemP256,
+            HpkeKemId::X25519HkdfSha256 => KemAlgorithm::DhKem25519,
+            _ => unreachable!("Unsupported KEM ({:?})", kem_id),
```

Comment:
Matching "_" for enumerated types is a bit of a footgun in my experience. Better to be explicit!

```suggestion
            NotImplemented(x) => unreachable!("Unsupported KEM ({:?})", x),
```

Reply:
done